File: Cap14CSharp/LiveProvider.cs
using System.Linq.Expressions;
using Cap14CSharp.Live.Search;
using LINQToLive;

namespace Cap14CSharp
{
    public class LiveProvider : QueryProvider
    {
        private string _liveKey;
        private SourceType _type;
        public LiveProvider(string liveKey, SourceType type)
        {
            _liveKey = liveKey;
            _type = type;
        }

        public override string GetQueryText(Expression expression)
        {
            return Translate(expression).TranslateToText();
        }

        public override object Execute(Expression expression)
        {
            var request = Translate(expression);
            var response = new LiveWebHelper().GetResultsFromLive(request);
            var items = response.Responses[0].Results;
            return new ParserResultados(items);
        }

        private SearchRequest Translate(Expression expression)
        {
            return new VisitanteArvore(_liveKey, _type).TraduzArvore(expression);
        }
    }
}

File: Cap14VB/LiveProvider.vb
Imports Cap14VB.Live.Search
Imports System.Linq.Expressions

Public Class LiveProvider
  Inherits QueryProvider
  Private _liveKey As String
  Private _type As SourceType
  Public Sub New(ByVal liveKey As String, ByVal type As SourceType)
    _liveKey = liveKey
    _type = type
  End Sub

  Public Overloads Overrides Function GetQueryText(ByVal expression As Expression) As String
    Return Translate(expression).TranslateToText()
  End Function

  Public Overloads Overrides Function Execute(ByVal expression As Expression) As Object
    Dim request = Translate(expression)
    Dim response = New LiveWebHelper().GetResultsFromLive(request)
    Dim items = response.Responses(0).Results
    Return New ParserResultados(items)
  End Function

  Private Function Translate(ByVal expression As Expression) As SearchRequest
    Return New VisitanteArvore(_liveKey, _type).TraduzArvore(expression)
  End Function
End Class
